### PR TITLE
Implement CSG tree simplification algorithms

### PR DESCRIPTION
**Pull request by sethrj**
This adds logic for replacing nodes in a CSG tree with a constant value, then propagating the implications of that down and up the tree. This is a substantial improvement of the simplification logic in SCALE, which still leaves unnecessary constants in the tree and cannot combine multiple nodes.

### CODE REVIEW

**Comment by sethrj** [comment]
@elliottbiondo Don't forget about this one, please!

File: src/orange/construct/detail/PostfixLogicBuilder.hh [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    //! \name Type aliases
+    using VecLogic = std::vector<logic_int>;
+    //!@}
+
+    static_assert(std::is_same_v<LocalSurfaceId::size_type, logic_int>,
+                  "unsupported: add enum logic conversion for different-sized "
+                  "face and surface ints");
+
+  public:
+    // Construct with reference to vector to append to
```
Comment (elliottbiondo) [inline src/orange/construct/detail/PostfixLogicBuilder.hh]: pointer

File: src/orange/construct/detail/PostfixLogicBuilder.hh [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    explicit inline PostfixLogicBuilder(CsgTree const& tree, VecLogic* logic);
+
+    //! Build from a node ID
+    inline void operator()(NodeId const& n);
+
+    // Visit an actual surface
```
Comment (elliottbiondo) [inline src/orange/construct/detail/PostfixLogicBuilder.hh]: actual?

File: src/orange/construct/detail/PostfixLogicBuilder.hh [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    inline void operator()(True const&);
```
Comment (elliottbiondo) [inline src/orange/construct/detail/PostfixLogicBuilder.hh:49]: Add doc strings

File: src/orange/construct/detail/PostfixLogicBuilder.hh [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+ */
+void PostfixLogicBuilder::operator()(Aliased const& n)
+{
+    // CELER_ASSERT_UNREACHABLE();
+    return (*this)(n.node);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * If replacing a queued node with a boolean, ours should match.
```
Comment (elliottbiondo) [inline src/orange/construct/detail/PostfixLogicBuilder.hh]: Add doc strings, current doc string as detail

File: src/orange/construct/detail/PostfixLogicBuilder.hh [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+ */
+void PostfixLogicBuilder::operator()(Negated const& n)
+{
+    (*this)(n.node);
+    logic_->push_back(logic::lnot);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * If replacing a queued node with a boolean, ours should match.
```
Comment (elliottbiondo) [inline src/orange/construct/detail/PostfixLogicBuilder.hh]: Same as above

File: src/orange/construct/detail/PostfixLogicBuilder.hh [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+#include "../CsgTree.hh"
+
+namespace celeritas
+{
+namespace csg
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Recursively construct a logic vector from a node with postfix operation.
```
Comment (elliottbiondo) [inline src/orange/construct/detail/PostfixLogicBuilder.hh:21]: This class could use some additional high-level documentation. While I that how the `ContainerVisitor` is calling the operator() functions, and that how `PostfixLogicBuilder::operator()(NodeId const& n)` provides recursion, this is pretty hard to follow.

File: src/orange/construct/detail/PostfixLogicBuilder.hh [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    CELER_EXPECT(logic_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Build from a node ID.
+ */
+void PostfixLogicBuilder::operator()(NodeId const& n)
+{
+    return this->visit_node_(*this, n);
```
Comment (elliottbiondo) [inline src/orange/construct/detail/PostfixLogicBuilder.hh]: Remove return here and in all other operator() functions

File: src/orange/construct/detail/NodeReplacementInserter.hh [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+class NodeReplacementInserter
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using VecNode = std::vector<std::pair<NodeId, Node>>;
+    //!@}
+
+  public:
+    // Construct with references to the stack and the replacement value
```
Comment (elliottbiondo) [inline src/orange/construct/detail/NodeReplacementInserter.hh]: pointer to the stack and reference to replacement value

File: src/orange/construct/detail/NodeReplacementInserter.hh [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    inline NodeReplacementInserter(VecNode* stack, Node const& repl);
+
+    // If replacing a queued node with a boolean, ours should match
```
Comment (elliottbiondo) [inline src/orange/construct/detail/NodeReplacementInserter.hh]: This comment doesn't make sense without the context that the function itself has a DBC check. I recommend adding standard doc strings to these functions.

File: src/orange/construct/detail/NodeReplacementInserter.hh [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    else
+    {
+        negated_ = Node{True{}};
+        join_token_ = op_or;  // any{...} = false
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * If replacing a queued node with a boolean, ours should match.
```
Comment (elliottbiondo) [inline src/orange/construct/detail/NodeReplacementInserter.hh]: Add standard doc string here and throughout

File: src/orange/construct/detail/NodeReplacementInserter.hh [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    Node const& repl_;
+    Node negated_;
+    OperatorToken join_token_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with references to the stack and the replacement value.
```
Comment (elliottbiondo) [inline src/orange/construct/detail/NodeReplacementInserter.hh]: pointer

**Review by elliottbiondo — CHANGES_REQUESTED** [review]

File: src/orange/construct/detail/PostfixLogicBuilder.hh (hunk 7 above)
Comment (sethrj) [reply src/orange/construct/detail/PostfixLogicBuilder.hh]: Sometimes this is a good thing to do (it ensures that the function you're calling actually returns `void` so you're not discarding the result), but you're right it doesn't add anything here.

File: src/orange/construct/detail/NodeReplacementInserter.hh (hunk 11 above)
Comment (sethrj) [reply src/orange/construct/detail/NodeReplacementInserter.hh]: In the doc strings I think pointer and reference can be effectively interchangeable. The type itself is clear from the code, the point of the documentation is to emphasize that you're referencing memory outside the class itself.

File: src/orange/construct/detail/PostfixLogicBuilder.hh (hunk 7 above)
Comment (elliottbiondo) [reply src/orange/construct/detail/PostfixLogicBuilder.hh]: The confusing part was really the inconsistency: some functions had returns and other didn't

File: src/orange/construct/detail/NodeReplacementInserter.hh (hunk 11 above)
Comment (elliottbiondo) [reply src/orange/construct/detail/NodeReplacementInserter.hh]: I agree that you shouldn't have to mention the type in a doc string, but if you are going to mention a type, it should be the correct type.

**Review by elliottbiondo — APPROVED** [review]
